<p><img src="https://files.cdn.thinkific.com/file_uploads/133292/images/a9e/c1a/bc3/P02.jpg"></p>

`MCSDSC02P01V3______`

`MCSDSC02P01V2______`

`MCSDSELIMC02P01V1010621`

<h3><font color="#FF0F00">Комментарий ассистента</font></h3>

Как обычно, заголовки (комментариев) будут трех цветов. <font color="89DD00">Зеленый</font> — все правильно (и если комментария нет, то это также значит, что все правильно). <font color="FFB800">Желтый</font> — можно доработать. <font color="FF0F00">Красный</font> — нужно исправить.

## Постановка задачи

На основании данных сервиса с тестового запуска на территории дизайн-завода Флакон и Хлебозавода №9 определить:  
* На какие кафе и в какие дни недели приходится больше всего заказов?
* Пользуются ли приложением чаще в обеденное время?
* В кафе приходят компаниями или поодиночке?  

Для этого:
1. Получите данные из файла bez_ocheredi.csv из папки data. Получите первые 10 строк таблицы и общую информацию о наборе. Опишите полученные данные.
2. Предобработайте данные: заполните пропущенные значения, удалите полные дубликаты, переименуйте заказы на правильное название Iskra Bar.
3. Посчитайте, на какие кафе и на какие дни недели приходятся больше всего заказов. Выведите результаты для Iskra Bar и Там Ем для понедельника, среды и пятница в таблице.
4. Напишите функцию для расчета количества заказов в обеденное время (с 12 до 16) и вечернее время (с 16 до 22)
5. Напишите функцию, которая определит моду (типичное значение) количества человек за столиком. Протестируйте функцию для Iskra Bar и Там Ем в обеденное и вечернее время.
6. Сделайте выводы.

## Импорт данных и библиотек

In [2]:
import pandas as pd

Прочитаем файл `bez_ocheredi.csv` и сохраним его в переменную `bo_data`.

In [3]:
bo_data = pd.read_csv('../../data/bez_ocheredi.csv')

Получим первые 10 строк таблицы.

In [4]:
bo_data.head(10)

userID    ресторан     weekday  ordered    Столик для 
0  b912c3c6a6ad   Iskra Bar     Monday     10:00             1
1  2dac0150717d   Iskra Bar     Monday     13:30             2
2  ba83aefab5dc   Iskra Bar    Tuesday     12:30             3
3  02996f1dd2ea   Iskra Bar  Wednesday     12:00             2
4  7c90be56d2ab   Iskra Bar   Thursday     14:30             3
5  51e0962387f7      Там Ем     Friday     12:30             2
6  c4f260a2b48a         NaN   Saturday     15:00             6
7  6b71309d6a8a  Крафтейнер     Monday     10:30             1
8  c5b969cb63a2  Бар Сибирь  Wednesday     15:30             2
9  b1570962e68c  Бар Сибирь  Wednesday     12:30             3

Получим общую информацию о данных из таблицы.

In [4]:
bo_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1024 entries, 0 to 1023
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   userID        1024 non-null   object
 1    ресторан     959 non-null    object
 2    weekday      1024 non-null   object
 3    ordered      1024 non-null   object
 4    Столик для   1024 non-null   int64 
dtypes: int64(1), object(4)
memory usage: 40.1+ KB


Столбцы в `bo_data` содержат следующую информацию:

* уникальный ID пользователя
* название ресторана
* день недели
* время заказа
* количество человек за столиком

### Выводы

* в исходных данных 1024 строки и 5 столбцов, содержащих данные с уникальным ID пользователя, название ресторана, день недели, время заказа и количество человек за столиком
* информация во всех столбцах, кроме названия ресторана, заполнена полностью
* название ресторана заполнено только в 959 строках
* в исходных данных в 4 столбцах тип данных - текст, в 1 столбце - целые числа

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Да, нужно указать на структуру данных, а также наличие пропусков. Все отлично.

## Предобработка данных

Получим названия столбцов.

In [5]:
bo_data.columns

Index(['userID ', ' ресторан ', ' weekday', ' ordered ', ' Столик для '], dtype='object')

Обращаем внимание, что помимо названий на разном языке, в них есть пробелы, которые затрудняют доступ к данным. Переименуем столбцы и проверим результат.

In [6]:
bo_data.set_axis(['user_id', 'restaurant', 'weekday', 'ordered_time', 'clients'], axis = 'columns', inplace = True)

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Отлично, но ordered можно назвать очевиднее. Например, ordered_time.  
<font color="#000080">***Поправила***

In [7]:
bo_data.columns

Index(['user_id', 'restaurant', 'weekday', 'ordered_time', 'clients'], dtype='object')

Еще раз проверим данные на наличие пропусков, просуммировав их.

In [8]:
bo_data.isnull().sum()

user_id          0
restaurant      65
weekday          0
ordered_time     0
clients          0
dtype: int64

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Эти методы равноценны, можно оставить любой из них.  
<font color="000080">***Поправила***

Менеджер проекта предупреждал, что некоторые заказы заводились на тестовую карточку кафе без названия, поэтому есть смысл заполнить такой пропуск словом admin.

In [9]:
bo_data['restaurant'] = bo_data['restaurant'].fillna('admin')

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Здесь лучше бы указать конкретный столбец, в котором хочешь заполнить пропуски `bo_data['restaurant']`.

<font color="000080">***Указала***

Проверяем, что пропусков не осталось.

In [10]:
bo_data['restaurant'].isnull().sum()

0

Установим количество явных дубликатов. Если найдутся, то удалим и снова проверим, что их не осталось.

In [11]:
bo_data.duplicated().sum()

24

In [12]:
bo_data = bo_data.drop_duplicates().reset_index(drop = True)

In [13]:
bo_data.duplicated().sum()

0

Скорее всего, такие полные дубликаты говорят о сбое при выгрузке данных. Необходимо сообщить об этом.

Еще одна проблема, о которой предупреждал менеджер проекта — дубликаты карточки Iskra Bar с неправильными названием бара - "Искра" и "Искр". На них заводились реальные заказы, поэтому необходимо вычислить их и переименовать на правильное название — Iskra Bar.

Объявим функцию `replace_restaurant()`, которая заменяет неправильное название, а потом проверяет успешность выполнения замены подсчетом (если замена прошла успешно — результат подсчета должен быть 0).

In [14]:
def replace_restaurant(name, newname):
    global bo_data
    bo_data = bo_data.replace(name, newname)
    check = bo_data[bo_data['restaurant'] == name].count().loc['user_id']
    return check

Передаем функции названия 'Искра' и 'Искр'.

In [15]:
replace_restaurant('Искр', 'Iskra Bar')

0

In [16]:
replace_restaurant('Искра', 'Iskra Bar')

0

### Выводы

* в представленном массиве данных было 65 пропусков в названии ресторанов. Пропуски заменены на 'admin'
* выявлено 24 явных дубликата. Они исключены из массива, но, похоже, была ошибка при выгрузке данных
* все карточки из Искра бара приведены к одному названию Iskra Bar. В исходных данных встречались названия 'Искр' и 'Искра'

## Распределение заказов в течение недели

Считаем, на какие кафе приходится больше всего заказов.

In [17]:
bo_data['restaurant'].value_counts()

Iskra Bar     287
Там Ем        200
Бар Сибирь    159
Северия       107
Крафтейнер     78
Коворкафе      66
Азия           62
admin          41
Name: restaurant, dtype: int64

Группируем по `weekday` и считаем, на какие дни приходится больше всего заказов.

In [18]:
bo_data['weekday'].value_counts()

Tuesday      251
Wednesday    242
Thursday     180
Monday       141
Friday       110
Saturday      52
Sunday        24
Name: weekday, dtype: int64

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Или можно еще метод `value_counts()` использовать, он еще и отсортирует по убыванию.  
<font color="000080">***Поменяла на value_counts***

Объявляем функцию, которая возвращает количество заказов для ресторана в определенный день.

In [19]:
def rest_weekday(dataset, rest, weekday):
    return dataset[(dataset['restaurant'] == rest) & (dataset['weekday'] == weekday)].count()['user_id']

Передадим ей заведения Iskra Bar и Там Ем — два флагманских места, по которым просили результаты прежде всего — и понедельник, среду и пятницу в качестве дня недели.

In [20]:
iskra_mon = rest_weekday(bo_data, 'Iskra Bar', 'Monday')
iskra_mon

48

In [21]:
iskra_wed = rest_weekday(bo_data, 'Iskra Bar', 'Wednesday')
iskra_wed

53

In [22]:
iskra_fri = rest_weekday(bo_data, 'Iskra Bar', 'Friday')
iskra_fri

29

In [23]:
tamem_mon = rest_weekday(bo_data, 'Там Ем', 'Monday')
tamem_mon

25

In [24]:
tamem_wed = rest_weekday(bo_data, 'Там Ем', 'Wednesday')
tamem_wed

52

In [25]:
tamem_fri = rest_weekday(bo_data, 'Там Ем', 'Friday')
tamem_fri

24

Сведем результат в таблицу.

In [26]:
table = bo_data.pivot_table(index = ['restaurant', 'weekday'], values = 'user_id', aggfunc = 'count')
table.query('weekday == ["Monday", "Wednesday", "Friday"] and restaurant == ["Iskra Bar", "Там Ем"]').transpose()

restaurant Iskra Bar                  Там Ем                 
weekday       Friday Monday Wednesday Friday Monday Wednesday
user_id           29     48        53     24     25        52

In [27]:
data = [
    ['Iskra Bar', iskra_mon, iskra_wed, iskra_fri],
    ['Там Ем', tamem_mon, tamem_wed, tamem_fri]
]
columns = ['Ресторан', 'Понедельник', 'Среда', 'Пятница']

table = pd.DataFrame(data=data, columns=columns)
table

Ресторан  Понедельник  Среда  Пятница
0  Iskra Bar           48     53       29
1     Там Ем           25     52       24

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Хорошо, что уже здесь используешь `pivot_table()`. Еще один из вариантов собрать таблицу — конструктор DataFrame:

```
data = [
    ['Iskra Bar', iskra_mon, iskra_wed, iskra_fri],
    ['Там Ем', tamem_mon, tamem_wed, tamem_fri]
        ]

columns = ['Ресторан', 'Понедельник', 'Среда', 'Пятница']

table = pd.DataFrame(data=data, columns=columns)
table
```

### Выводы

* больше всего заказов приходится на Искра Бар и Там Ем
* итого по всем ресторанам больше всего заказов приходится на вторник, среду, четверг
* меньше всего заказов в выходные - видимо, заранее заказывают в ресторанах именно в будни люди, которые работают неподалёку и не могут тратить много времени на ожидание ресторана/еды
* если анализировать 3 будних дня (пн, ср, птн) в Искра баре и Там Ем, меньше всего заказов в пятницу, больше всего - в среду

## Распределение заказов в течение дня

Сравним количества заказов, сделанных в обеденное и вечернее время — с 12 до 16 и с 16 до 22. Для этого объявим функцию, которая принимает данные и 2 аргумента времени.

In [28]:
# преобразовать текстовое написание времени заказа столика в другой тип данных
bo_data['hour'] = pd.to_datetime(bo_data['ordered_time']).dt.hour.apply(float)
bo_data.head() # проверка

user_id restaurant    weekday ordered_time  clients  hour
0  b912c3c6a6ad  Iskra Bar     Monday        10:00        1  10.0
1  2dac0150717d  Iskra Bar     Monday        13:30        2  13.0
2  ba83aefab5dc  Iskra Bar    Tuesday        12:30        3  12.0
3  02996f1dd2ea  Iskra Bar  Wednesday        12:00        2  12.0
4  7c90be56d2ab  Iskra Bar   Thursday        14:30        3  14.0

In [29]:
#НАПИСАННОЕ В ЭТОЙ ЯЧЕЙКЕ НЕ ИСПОЛЬЗУЕТСЯ ДАЛЬШЕ В РАСЧЁТАХ В ЭТОМ БЛОКЕ (ТОЛЬКО В СЛЕДУЮЩЕМ)

# функция по определению интервала
def set_interval(hour):
    if hour >= 12:
        if hour < 16:
            return '12-16'
        else:
            return '16-22'
    else:
        return '10-12'
# добавить новый столбец interval с интервалом
bo_data['interval'] = bo_data['hour'].apply(lambda hour: set_interval(hour))
bo_data.head() # проверка

user_id restaurant    weekday ordered_time  clients  hour interval
0  b912c3c6a6ad  Iskra Bar     Monday        10:00        1  10.0    10-12
1  2dac0150717d  Iskra Bar     Monday        13:30        2  13.0    12-16
2  ba83aefab5dc  Iskra Bar    Tuesday        12:30        3  12.0    12-16
3  02996f1dd2ea  Iskra Bar  Wednesday        12:00        2  12.0    12-16
4  7c90be56d2ab  Iskra Bar   Thursday        14:30        3  14.0    12-16

In [30]:
# написать функцию расчета кол-ва заказов в разные временные интервалы
def count_interval(dataset, time_start, time_finish):
    return dataset[(dataset['hour'] >= time_start) & (dataset['hour'] < time_finish)].count().loc['user_id']

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

В этом этапе можно обойтись и без преобразования и передавать сами строки:

```
def timing(data, from_time, to_time):
    return data[(data['ordered_time'] >= from_time) & (data['ordered_time'] < to_time)]['ordered_time'].count()

timing(bo_data, '12:00', '16:00')
   
```
<font color="000080">***сделала ниже вариант2***

Передадим промежуток с 12:00 и 16:00

In [31]:
count_interval(bo_data, 12, 16)

543

И с 16:00 до 22:00

In [32]:
count_interval(bo_data, 16, 22)

338

In [37]:
# вариант 2
def count_int2(dataset, time_start, time_finish):
    return dataset[(dataset['ordered_time'] >= time_start) & (dataset['ordered_time'] < time_finish)]['ordered_time'].count()
count_int2(bo_data, '16:00', '22:00')

338

### Выводы

* гипотеза о том, что больше всего заказов приходится на обеденное время, верная: приложение нужно в первую очередь во время обеда, чтобы сэкономить время на ожидание официанта/заказ

## Типичное количество человек за столиком

Выясним, для какого количества человек обычно заказывают столик в разное время на примере Iskra Bar и Там Ем. Для этого напишем функцию, которая принимает данные, название ресторана, 2 времени и считает моду.

In [38]:
def count_moda(rest, start, finish):
    global bo_data
    moda_clients = bo_data.drop(['user_id', 'weekday', 'ordered_time'], axis = 1)
    md = moda_clients[(moda_clients['restaurant'] == rest) & (moda_clients['hour'] >= start) & (moda_clients['hour'] < finish)].mode()
    return md[['restaurant', 'clients', 'interval']]

In [39]:
# "вручную" проверяла корректность работы формулы на конкретных примерах
for i in bo_data['clients'].unique():
    print(f'{i}: {bo_data[(bo_data["restaurant"] == "Там Ем") & (bo_data["clients"] == i) & (bo_data["interval"] == "12-16")].count().loc["user_id"]}')

1: 35
2: 43
3: 15
6: 1
4: 3
5: 0


<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Круто, ты считаешь еще и количество каждого значения, но в упрощенной версии можно просто в одну строчку написать логическое условие и прицепить к нему метод `mode()`.

```
def mode_of_people(data, restaurant_name, from_time, to_time):
    return data[(data['ordered_time'] >= from_time) & (data['ordered_time'] < to_time) & (data['restaurant'] == restaurant_name)]['table_for'].mode()
```  
<font color="000080">***сделала ниже вариант2***

Проверим функцию на двух ресторанах, а также дневном и вечернем промежутке времени.

In [40]:
count_moda('Iskra Bar', 12, 16)

restaurant  clients interval
0  Iskra Bar        2    12-16

In [44]:
count_moda('Iskra Bar', 16, 22)

restaurant  clients interval
0  Iskra Bar      1.0    16-22
1        NaN      NaN      NaN

In [45]:
count_moda('Там Ем', 12, 16)

restaurant  clients interval
0     Там Ем        2    12-16

In [46]:
count_moda('Там Ем', 16, 22)

restaurant  clients interval
0     Там Ем        1    16-22

In [48]:
# вариант 2
def count_moda2(dataset, rest, start, finish):
    return dataset[(dataset['restaurant'] == rest) & (dataset['ordered_time'] >= start) & (dataset['ordered_time'] < finish)]['clients'].mode()
count_moda2(bo_data, 'Там Ем', '12:00', '16:00')

0    2
dtype: int64

### Выводы

* в обеденное время заказывают чаще на двоих - видимо, чаще ходят обедать вместе с коллегами, а не поодиночке
* в вечернее время чаще столик на одного - возможно, чтобы быстрее поесть перед тренировкой или забрать ужин с собой на вынос

## Общие выводы

* гипотеза о том, что приложением чаще пользуются в обед, подтверждается представленными данными
* в обеденное время чаще всего по приложению делают заказ на двоих, в то время как вечером заказ чаще на одного
* больше всего заказов по приложению приходится на середину рабочей недели - вторник, среда, четверг. В выходные количество заказов существенно падает - посетители приходят посидеть/пообщаться, нет необходимости делать заказ заранее
* есть вопросы к качеству выгруженных данных - были пропуски, некорректно заполненные названия ресторана и дубликаты

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Отлично, в конце сверяемся на все ли вопросы менеджера даны ответы.